File: Cpp/nhapmonBFS.cpp
#include <vector>
#include <queue>
#include <iostream>
#include <tuple>

typedef std::vector<int> vectori;

bool is_valid(int x, int y, int n, int m) {
    return 0 <= x && x < n && 0 <= y && y < m;
}

int solve(int n, int m, int x1, int y1, int x2, int y2) {
    vectori dx = {-2, -1, 1, 2, 2, 1, -1, -2};
    vectori dy = {1, 2, 2, 1, -1, -2, -2, -1};
    std::vector<std::vector<bool>> is_visited(n, std::vector<bool>(m));
    std::queue<std::tuple<int, int, int>> q;
    q.push(std::make_tuple(x1, y1, 0));
    is_visited[x1][y1] = true;

    while (!q.empty()) {
        std::tuple<int, int, int> front = q.front();
        q.pop();
        int x = std::get<0>(front);
        int y = std::get<1>(front);
        int steps = std::get<2>(front);
        if (x == x2 && y == y2) {
            return steps;
        }
        for (int i = 0; i < 8; i++) {
            int nx = x + dx[i], ny = y + dy[i];
            if (is_valid(nx, ny, n, m) && !is_visited[nx][ny]) {
                q.push(std::make_tuple(nx, ny, steps + 1));
                is_visited[nx][ny] = true;
            }
        }
    }
    return -1;
}

int main() {
    std::ios_base::sync_with_stdio(false);
    std::cin.tie();
    std::cout.tie();
    int n, m, x1, y1, x2, y2;
    std::cin >> n >> m >> x1 >> y1 >> x2 >> y2;
    std::cout << solve(n, m, x1, y1, x2, y2) << std::endl;
    return 0;
}


// code by duck it1

File: Python/nhapmonBFS.py
from collections import deque

def is_valid(x, y, n, m):
    return 0 <= x < n and 0 <= y < m

def solve(n, m, x1, y1, x2, y2):
    dx = [-2, -1, 1, 2, 2, 1, -1, -2]
    dy = [1, 2, 2, 1, -1, -2, -2, -1]

    is_visited = [[False] * m for _ in range(n)]
    queue = deque([(x1, y1, 0)])
    is_visited[x1][y1] = True
    while queue:
        x, y, steps = queue.popleft()
        if x == x2 and y == y2:
            return steps
        for i in range(8):
            nx, ny = x + dx[i], y + dy[i]
            if is_valid(nx, ny, n, m) and not is_visited[nx][ny]:
                queue.append((nx, ny, steps + 1))
                is_visited[nx][ny] = True
    return -1

n, m = map(int, input().split())
x1, y1 = map(int, input().split())
x2, y2 = map(int, input().split())
print(solve(n, m, x1, y1, x2, y2))
